### PR TITLE
`Exam mode`: Allow instructors to view unsubmitted exams and adjust summary layout

### PR DESCRIPTION
**Pull request by florian-glombik**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
I ran into this error while working on https://github.com/ls1intum/Artemis/issues/7561 and think that it is not good for the UX to navigate to a page that is not loaded properly.
The underlying issue is that there is no access granted to the grade.

### Description
* Allowing instructors to access the grade
* Improving existing check for test runs
* Adding clear indicator when an exam is viewed that was not submitted
* Adjusted the layout of the exam summary
* Fixing teamscale test that fails on develop
* Adjusted test for grade access control

~Disabling the button that forwards to the student exam summary instead of forwarding to a broken page as the server endpoint denies access to the grade of an unsubmitted exam.~

~This issue could also be resolved by allowing access to the grade for instructors - since there is the explicit error message `You are not allowed to access the grade summary of a student exam which was NOT submitted!` thrown I assumed that there are some explicit thoughts behind denying this view for instructor.
As I do not know a scenario where the instructor would like to view unsubmitted exams, I think that just disabling the button that forwards to the student exam is a better solution than letting the instructor run into error messages.~

### Steps for Testing
- 1 Student
- 1 Exam that starts in the future _(with prepared exams where the student is registered)_
- 1 Instructor
- 1 Exam with an unsubmitted student exam

1. `student`: verify the preview page of an exam does still have the general information centered
2. `instructor`: assess the exam
3. `instructor`: release the grades (= _Release Dates of Results_ in the past)
4. `instructor`: try to view the unsubmitted Student Exam
5. `instructor`: see that no error message is displayed, but instead a banner that the currently viewed exam was not submitted (verify translation for English and German)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------------:|
| [exam-result-summary.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5732/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/summary/exam-result-summary.component.ts.html) | 91.09% |                           ✅ |
| [exam-result-overview.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5732/latest/artifact/shared/Coverage-Report-Client-Tests/app/exam/participate/summary/result-overview/exam-result-overview.component.ts.html) | 85.71% |                          ✅ |

#### Server

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------------:|
| [ExamService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5732/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.exam/ExamService.html) | 93% |                           ✅ |
| [StudentExamResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS5732/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest/StudentExamResource.html) | 86% |                          ✅ |

#### New
new Layout
image

banner instead of a notification when viewing an resubmitted exam
image

#### Old
https://github.com/ls1intum/Artemis/assets/63976129/2f044099-95f6-4712-9574-46ef7d112eb4

## Summary by CodeRabbit

- **New Features**
  - Introduced collapsible card components for exam content.
  - Enhanced exam result summary pages with new collapsible functionality and improved layout.
  - Added new methods for fetching and displaying Iris course and exercise settings.

- **Enhancements**
  - Updated exam score calculation methods to accommodate bonus points.
  - Improved student exam grades summary with additional access control logic.
  - Redesigned exam general information and result overview pages for better readability and user experience.
  - Expanded Cypress test support with increased timeout for starting complaints.

- **Bug Fixes**
  - Fixed UI inconsistencies in exercise card headers and grading key tables.
  - Corrected margin and alignment issues in general information components.

- **Refactor**
  - Streamlined access checks and renamed methods for clarity in exam services.
  - Updated unit tests to reflect changes in collapsible card components and exam result calculations.

- **Style**
  - Adjusted styles for tables and collapsible components to enhance visual presentation.

- **Tests**
  - Added new unit tests for collapsible card functionality.
  - Removed outdated test cases related to removed functionalities.
  - Refactored existing tests to align with updated component structures.

### CODE REVIEW

File: src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.html [hunk 1]
```diff
@@ -2,125 +2,127 @@
...
-                                      achievedBonus: studentExamWithGrade.studentResult.gradeWithBonus.bonusGrade,
-                                      bonusFromTitle: studentExamWithGrade.studentResult.gradeWithBonus.bonusFromTitle
-                                  }
-                    }}
-                </h5>
+    <div class="exam-points-summary-container" [class.d-flex]="displayGradeKeysBesideTable">
+        <div class="grade-results-table-container">
+            <table class="table table-striped table-hover w-auto" id="result-overview-table">
+                <thead>
+                    <tr class="text-nowrap">
```
Comment (florian-glombik) [inline src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.html:9]: This part was changed a bit (wrapped in a container soo the grading keys can be displayed at the right side if enough space)

File: src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.html [hunk 2]
```diff
@@ -2,125 +2,127 @@
...
+                        </th>
+                        <th *ngIf="studentExamWithGrade?.maxBonusPoints">
+                            {{ studentExamWithGrade?.maxBonusPoints }}
+                        </th>
+                    </tr>
+                </tfoot>
+            </table>
+            <div class="d-flex align-items-center">
+                <div class="you-achieved-container">
+                    <div *ngIf="studentExamWithGrade?.maxBonusPoints" class="h5">
```
Comment (florian-glombik) [inline src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.html:92]: Changed the `h5` to only styling, to improve the metadata structure at least a bit

File: src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.html [hunk 3]
```diff
@@ -2,125 +2,127 @@
...
+                        <td class="text-center" *ngIf="showIncludedInScoreColumn">{{ exerciseService.isIncludedInScore(exercise) }}</td>
+                        <td class="text-center">
+                            {{ studentExamWithGrade?.achievedPointsPerExercise?.[exercise.id!] ?? 0 }}
+                        </td>
+                        <td class="text-center">
+                            <div *ngIf="exercise.maxPoints != undefined">
+                                {{ exercise.includedInOverallScore === IncludedInOverallScore.INCLUDED_AS_BONUS ? 0 : exercise.maxPoints }}
+                            </div>
+                        </td>
+                        <td class="text-center">
```
Comment (florian-glombik) [inline src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.html:47]: centered the numbers in the table

**Review by mateusmm01 — DISMISSED** [review]
Demonstrated during testing session. Worked as expected.

**Review by laurenzfb — DISMISSED** [review]
Demonstrated in testing session on ts9. Worked as expected!

**Review by sarpsahinalp — APPROVED** [review]
Demonstrated during testing session. Worked as expected.

**Review by vinceclifford — APPROVED** [review]
Works as expected, demonstrated during testing session

**Review by muradium — APPROVED** [review]
Demonstrated during test session, works as expected

File: src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.ts [hunk 4]
```diff
@@ -89,11 +89,58 @@ export class ExamResultOverviewComponent implements OnInit, OnChanges {
...
+        this.overallAchievedPercentageRoundedByCourseSettings = this.getOverallAchievedPercentageRoundedByCourseSettings();
+    }
+
+    /**
+     * used as fallback if not pre-calculated by the server
+     */
+    private sumExerciseScores() {
+        let exerciseScoreSum = 0;
+
+        this.studentExamWithGrade.studentExam?.exercises?.forEach((exercise) => {
```
Comment (reschandreas) [inline src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.ts:102]: maybe .reduce would be a better fit here

**Review by reschandreas — CHANGES_REQUESTED** [review]
overall code lgtm, one comment

File: src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.ts (hunk 4 above)
Comment (florian-glombik) [reply src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.ts]: Addressed with https://github.com/ls1intum/Artemis/pull/7588/commits/422f4aa5665a45f1f6e0254ffb16cedca492b201

**Review by milljoniaer — APPROVED** [review]
Code looks ok. LGTM 🚀

**Review by milljoniaer — DISMISSED** [review]
Approve e2e fix changes

**Review by mateusmm01 — DISMISSED** [review]
Tested on TS1, works as expected!

**Review by reschandreas — APPROVED** [review]
overall code lgtm

**Review by reschandreas — APPROVED** [review]
reapprove after fixing merge conflicts

File: src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.ts [hunk 5]
```diff
@@ -175,4 +227,8 @@
     toggleBonusGradingKey(): void {
         this.isBonusGradingKeyCollapsed = !this.isBonusGradingKeyCollapsed;
     }
+
+    toggleCollapse(resultOverviewSection: ResultOverviewSection) {
+        return () => (this.isCollapsed[resultOverviewSection] = !this.isCollapsed[resultOverviewSection]);
+    }
```
[the remark this answers is not in the document]
Comment (florian-glombik) [reply src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.ts:233]: I am not aware of another way to keep the component reusable

File: src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.ts [hunk 6]
```diff
@@ -89,11 +97,55 @@ export class ExamResultOverviewComponent implements OnInit, OnChanges {
...
+        });
+
+        if (numberOfExercises === 0) {
+            return 0;
+        }
+
+        return summedPercentages / numberOfExercises;
     }
 
     /**
```
[the remark this answers is not in the document]
Comment (florian-glombik) [reply src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.ts:151]: See src/test/javascript/spec/component/exam/participate/summary/result-overview/exam-result-overview.component.spec.ts

File: src/main/webapp/app/exam/participate/summary/collapsible-card.component.html [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+        <ng-content select=".header"></ng-content>
+        <button class="btn rotate-icon" [class.rotated]="!isCardContentCollapsed" [attr.aria-expanded]="isCardContentCollapsed" [attr.aria-controls]="isCardContentCollapsed">
+            <fa-icon size="2x" [icon]="faAngleRight"></fa-icon>
+        </button>
+    </div>
+
+    <div class="card-body question-card-body" [ngbCollapse]="isCardContentCollapsed">
+        <ng-content select=".collapsible-content"></ng-content>
+    </div>
+</div>
```
[the remark this answers is not in the document]
Comment (florian-glombik) [reply src/main/webapp/app/exam/participate/summary/collapsible-card.component.html]: In general yes, screen reader support in Artemis requires more work and is something that I think should be looked at seperately.

File: src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.html [hunk 8]
```diff
@@ -3,103 +3,110 @@
...
+                                        <span [class]="exerciseInfos[exercise.id!].colorClass"> {{ exerciseInfos[exercise.id!].achievedPercentage }} % </span>
+                                    } @else {
+                                        -
+                                    }
+                                </td>
+                                @if (studentExamWithGrade?.maxBonusPoints) {
+                                    <td>
+                                        @if (exercise.bonusPoints != undefined) {
+                                            <span>{{
+                                                exercise.includedInOverallScore === IncludedInOverallScore.INCLUDED_AS_BONUS ? exercise.maxPoints : exercise.bonusPoints
```
[the remark this answers is not in the document]
Comment (florian-glombik) [reply src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.html:72]: As far is I can see this statement is wrong.

See line 52 src/main/webapp/app/exam/participate/summary/result-overview/exam-result-overview.component.ts

**Review by milljoniaer — APPROVED** [review]
Reapprove after code changes

**Review by konrad2002 — APPROVED** [review]
manually tested on legacy ts11, works as expected

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java [hunk 9]
```diff
@@ -424,7 +424,8 @@ private Map<Long, BonusSourceResultDTO> calculateExamScoresAsBonusSource(Long ex
             StudentExam studentExam = studentExamRepository.findWithExercisesByUserIdAndExamId(targetUser.getId(), examId, IS_TEST_RUN)
                     .orElseThrow(() -> new EntityNotFoundException("No student exam found for examId " + examId + " and userId " + studentId));
 
-            StudentExamWithGradeDTO studentExamWithGradeDTO = getStudentExamGradesForSummaryAsStudent(targetUser, studentExam, IS_TEST_RUN);
+            StudentExamWithGradeDTO studentExamWithGradeDTO = getStudentExamGradesForSummary(targetUser, studentExam,
+                    authorizationCheckService.isAtLeastInstructorInCourse(studentExam.getExam().getCourse(), targetUser));
             var studentResult = studentExamWithGradeDTO.studentResult();
             return Map.of(studentId, new BonusSourceResultDTO(studentResult.overallPointsAchieved(), studentResult.mostSeverePlagiarismVerdict(), null, null,
                     Boolean.TRUE.equals(studentResult.submitted())));
```
[the remark this answers is not in the document]
Comment (florian-glombik) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamService.java:431]: Not an issue as the grade as not relevant, as far as I know, if an exam was not submitted. Furthermore this seems to be out of scope for me

**Review by laurenzfb — APPROVED** [review]
tested on ts3. works as described

**Pull request merged**